### PR TITLE
make nhwc transformer only apply to CPU ep. 

### PR DESCRIPTION
**Pull request by Jamather**
**Description**: 
Also clean up nchwc transformer to be more consistent with other transformers (taking an argument of supported EPs instead of hardcoding the behavior)

**Motivation and Context**
- QLinearConv does not work with DML EP because this optimizer intended for CPU EP is wrongfully applied to it.

### CODE REVIEW

**Review by fdwr — APPROVED** [review]
Some style inconsistencies (I'd just mimic `BiasDropoutFusion`), but functionally looks good to me.

**Comment by fdwr** [comment]
Adding Scott as Tom is on a different team now.

File: onnxruntime/core/optimizer/nhwc_transformer.h
Comment (skottmckay) [inline onnxruntime/core/optimizer/nhwc_transformer.h:26]: nit: alignment looks off. please use clang-format.

File: onnxruntime/core/optimizer/graph_transformer_utils.cc
Comment (skottmckay) [inline onnxruntime/core/optimizer/graph_transformer_utils.cc:347]: This is a good example of why it's better for the EP to do its own optimizations. The core ORT code should not need to be aware of any non-ORT EP requirements as that doesn't scale.

**Review by skottmckay — DISMISSED** [review]
:shipit:

**Review by fdwr — DISMISSED** [review]

File: onnxruntime/core/optimizer/graph_transformer_utils.cc
Comment (yufenglee) [reply onnxruntime/core/optimizer/graph_transformer_utils.cc:348]: Level 2 is EP dependent optimization. Level 3 is for CPU EP ARCH dependent optimization. That's why EP is hardcoded in the transformer.

I just traced the code history. NhwcTransformer also had CPU EP check previously, but was removed in this PR:
[https://github.com/microsoft/onnxruntime/pull/9284/files#diff-930bf0c155479c9a2e17b4996b0ae133e7db12e3161b6ac868e6e9756d76771bL486-L488](https://github.com/microsoft/onnxruntime/pull/9284/files#diff-930bf0c155479c9a2e17b4996b0ae133e7db12e3161b6ac868e6e9756d76771bL486-L488)

**Comment by fdwr** [comment]
Seems the Linux build machines are running out of space [`No space left on device`](https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=672959&view=logs&j=7536d2cd-87d4-54fe-4891-bfbbf2741d83&t=c97f7c87-5539-44f2-8270-d1646a3bf305). Retrying it again, in case fluke.

**Review by fdwr — APPROVED** [review]

**Comment by skottmckay** [comment]
Was the change to cmake/external/onnx intentional?

**Comment by fdwr** [comment]
Seems okay to me, but will defer to @skottmckay. Glad the pipelines finally passed!

**Comment by skottmckay** [comment]
Sorry - taking a step back and considering @yufenglee's comment, would it be simpler just to re-instate the old hardcoding that the level 3 transformers are for the CPU EP only (along with the change to graph_partitioner.cc so remaining nodes are assigned to the CPU EP)? If there's no intention for the level 3 transformers to ever be applicable to other EPs it's slightly misleading to configure it with a list of supported EPs. 

Also note that Ed's PR (#12791) will remove the need for the change in graph_partitioner.cc very shortly.

**Comment by Jamather** [comment]
There are other transformers that only apply to the CPU EP, but follow the same pattern. Isn't this more consistent with the rest of the codebase?

**Comment by Jamather** [comment]
considering the other Level 3 transformer reads
transformers.emplace_back(std::make_unique<ConvAddActivationFusion>(cpu_ep));
I think this is still more consistent

**Comment by skottmckay** [comment]
Consistently following an incorrect pattern isn't necessarily the best choice either though. :P

Level 1: no EP list (as these use ONNX ops and run pre-partitioning so nodes have not been assigned to EPs)
Level 2: EP list (fwiw this isn't the best setup as EP specific logic bleeds into ORT core code - the EP should really have an easy way to do these fusions directly during partitioning, and if we provide that the EP list here would go away).
Level 3: was no EP, but this PR changes it to take an EP.

'consistent' is also subjective.  

If the level 3 transformers are only ever intended for usage with the ORT CPU EP, their implementations can be hardcoded to only process nodes assigned to that, and the ctor doesn't need to take a list of EPs. That more clearly advertises that the EPs it works with is not configurable. It's also less code change here.

Adding EP lists to level 3 transformers is also an option, but that's more code change, and potentially EP lists will go away in the future. 

@yufenglee do you have a preference? 

Unrelated to your changes, but now that I look at it again I think the additional level 3 optimizer for Conv+Add+Activitation should be in level 2 and either it should be disabled if we're going to add the NCHWc transformer or the NCHWc transformer should be updated to expect Conv+Add+Activation fusion ran. Putting it in level 3 pollutes what level 3 means.

**Comment by fdwr** [comment]
> Consistently following an incorrect pattern isn't necessarily the best choice either though. :P

But also, perfect is the enemy of the good. If there's a broader effort to redesign transformers all at once, that's cool, and that can happen holistically with a complete view in that other PR. Can we get this bug in though to unblock us, and then Yufeng or Scott can decide how they want to change the meaning of level 3 and ep lists?

**Comment by Jamather** [comment]
I'll revert it to the hardcoded check for CPU ep.

**Review by skottmckay — APPROVED** [review]
:shipit:

**Comment by skottmckay** [comment]
@edgchen1 is this is merged before your PR just drop the changes in inference_session.cs and graph_partitioner.cc when merging from main as I don't think they should be necessary with the new approach.

**Pull request merged**
